snapshot/restore/local_fs_output.go

```diff
@@ -326,6 +336,81 @@ func write(targetPath string, r fs.Reader) error {
 	return nil
 }
 
+type sparseWriter struct {
+	f *os.File
+}
+
+// nolint:wrapcheck
+func (sw *sparseWriter) Write(b []byte) (int, error) {
+	w := 0
+
+	bSize, err := stat.GetFileAllocSize(sw.f.Name())
+	if err != nil {
+		return w, err // nolint:wrapcheck
+	}
+
+	for i, j := 0, bSize; i < len(b); i, j = int(j), j+bSize {
+		if j > uint64(len(b)) {
+			j = uint64(len(b))
+		}
+
+		isZeros := true
+		block := b[i:j]
+
+		for _, cb := range block {
+			if cb != 0 {
+				isZeros = false
+				break
+			}
+		}
+
+		if isZeros { // nolint:nestif
+			head, err := sw.f.Seek(0, 1)
+			if err != nil {
+				return w, err
+			}
+
+			fi, err := sw.f.Stat()
+			if err != nil {
+				return w, err
+			}
+
+			// If at the end of file, extend it
+			if head == fi.Size() {
+				head += int64(len(block))
+
+				err = sw.f.Truncate(head)
```

Comment:
Can we call Truncate() once at the very beginning with the final size - this will enable us to parallelize the work and speed id up, which for sparse files (which can be up to hundreds of gigabytes) is going to be critical.

We can:

1. Create result file.
2. Call Truncate on it to the requested size (which we know up front).
3. Divide the work into large segments, say 1GiB each (power of two is likely important here).
   [0..1GiB), [1GiB..2GiB), [2GiB..3GiB), [3GiB..4GiB) .... 
4. Create a gorotuine that will be returning these ranges in a channel.
5. Start N goroutines that will be reading from the channel, and:
       - seek to the start offset in the input and output
       - processing N bytes up to the provided count
6. Wait for all goroutines to finish work.
7. Close the result file.

Reply:
That sounds really good actually–thanks for listing out the steps. I can follow up this PR with another one.